diff_drive_controller/src/speed_limiter.cpp

```diff
@@ -101,10 +105,43 @@ namespace diff_drive_controller
 
     if (has_acceleration_limits)
     {
-      const double dv_min = min_acceleration * dt;
-      const double dv_max = max_acceleration * dt;
-
-      const double dv = clamp(v - v0, dv_min, dv_max);
+      double dv_min, dv_max, dv;
+
+      if (v0 >= 0.0 && v >= 0.0)
+      {
+        // Moving in positive direction
+        dv_min = min_deceleration * dt;
+        dv_max = max_acceleration * dt;
+        dv = clamp(v - v0, dv_min, dv_max);
+      }
+      else if (v0 <= 0.0 && v <= 0.0)
+      {
+        // Moving in the negative direction
+        dv_min = min_acceleration * dt;
+        dv_max = max_deceleration * dt;
+        dv = clamp(v - v0, dv_min, dv_max);
+      }
+      else
+      {
+        // Transitioning from positive to negative velocity
+        if (v0 > 0)
+        {
+          dv = min_deceleration * dt;
+          if (v0 + dv < 0)
+          {
+            dv = min_acceleration * (dt - v0 / min_deceleration);
```

Comment:
I readjusted it so now I calculate dv_min and dv_max and clamp in between that like how the general case does it.

Has an added benefit of if the decelerations were poorly defined (i.e. a positive min deceleration), we will simply stay at the last velocity, instead of decelerating in the wrong direction and increasing the velocity error.